Fix information schema queries when cluster state has outdated views

BaurzhanSakhariev opened the pull request:
Relates to https://github.com/crate/crate/issues/14377

BaurzhanSakhariev commented on server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:
```diff
@@ -79,13 +78,10 @@ public ViewInfo create(RelationName ident, ClusterState state) {
-                throw e;
-            }
+            // Statement could not be analyzed, because the referenced table either not found
+            // or has been updated and view definition became incompatible with the new schema (https://github.com/crate/crate/issues/14377).
+            columns = Collections.emptyList();
+
```
Note, that information schema would be working and "corrupted" view won't be displayed. 

Drawback of such solution (including ignoring `RelationUnknown` exception on master) is having unusable views in the cluster state.  `select * from outdated_view` would raise the same casting error. 

Another ways to fix it:

1. Keep views in sync with tables. I'm not sure it's possible and practical in all scenarios. In the reported issue user wanted to subtract an interval from a timestamp and expected to get a timestamp. We can force view to return array with a single value having the "expected" result but such view can, in turn, break other user queries, dependant on that view.

2.  Ignore all exceptions but also drop the corrupted view as it's anyway unusable. It can also break user's dependant queries (also changing cluster state when accessing information schema is not really safe).

3. Ignore exceptions but maybe log a warning on information schemas queries (view X is outdated and cannot be accessed) and show a better error message on `select * from outdated_view`?

proddata replied:
> Drawback of such solution (including ignoring RelationUnknown exception on master) is having unusable views in the cluster state.

If it can be fixed e.g. by `CREATE OR REPLACE VIEW ....` I think that is fine.

ad 1) I think that is also bad in table replacement scenarios that we rely on, as reindexing is sometimes needed
ad 2) see 1

ad 3) Instead of a log message (which is limited to host admins) I would rather have an error message in `information_schema.views`

BaurzhanSakhariev replied:
> have an error message in information_schema.views

had a concern that PG doesn't have anything like that but we already have our custom `owner` column and in general custom column should be fine in terms of compatibility. Sounds fine to  me, @matriv, what do you think?

matriv replied:
I think that this PR's solution is the way to go. The view is erroneous when you try to query it and then you can still see what is the query stored from the view through `information_schema.views` and `pg_views` and fix it.

- I wouldn't add extra cols to `information_schema.views` which is SQL standard.
- Adding a log message when querying those catalog info tables, I think it's not helpful for the user.

I have another suggestion, how about adding a comment like:
``-- Corrupeted - needs fix --`` on top of the query of the view, returned in those tables?

BaurzhanSakhariev replied:
> I have another suggestion, how about adding a comment like:
-- Corrupted - needs fix -- on top of the query of the view, returned in those tables?

Will do, thanks! Should we mention a better error message on broken queries in the entry?

BaurzhanSakhariev replied:
Thanks, added a hint and updated a change entry.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/ViewsITest.java:
```diff
@@ -55,6 +58,36 @@ public void dropViews() {
+        // and view definition has "corrupted, needs fix" hint as a comment.
+
+        execute("SELECT view_definition FROM information_schema.views where table_name = 'v01'");
+        assertThat(response).hasRows(
+            "/* Corrupted view, needs fix */ SELECT CAST(\"ts\" AS timestamp) - CAST('1 hour' AS interval)",
+            "FROM \"t01\""
+        );
```
I probably misunderstand syntax of hasRows params but to me it looks like we expect 2 rows. But in runtime it's a single row with \n before FROM.

Found an [existing example](https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/integrationtests/ViewsITest.java#L70-L73) where we do such assertion, also use multiple rows, probably it's fine.

BaurzhanSakhariev commented on server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:
```diff
@@ -79,14 +80,12 @@ public ViewInfo create(RelationName ident, ClusterState state) {
+            analyzeError = true;
         }
-        return new ViewInfo(ident, view.stmt(), columns, view.owner());
+        String viewDefinition = analyzeError ? String.format(Locale.ENGLISH, "/* Corrupted view, needs fix */ %s", view.stmt()) : view.stmt();
```
Didn't use -- as it makes comments the whole statement, I tried locally enclosing with `-- comment -- SQL` and and didn't work.

matriv replied:
Because you'd need a `\n` in between the comment and the query.

matriv commented on docs/appendices/release-notes/5.3.5.rst:
```diff
@@ -80,3 +80,8 @@ Fixes
   leave behind empty partitions if ``NULL`` or ``CHECK`` constraint on
   partitioned by column failed.
 
+- Fixed an issue which caused errors on querying ``information_schema`` tables
```
I would rephrase it a bit to make it more generic:
```
Fixed an issue which caused errors on querying ``information_schema`` tables
when the query of a ``VIEW`` is erroneous due to changes made to the underlying tables/views. Also, added a comment to view definition in ``pg_catalog.pg_views`` and ``information_schema.views`` tables to denote that a ``VIEW``'s query is erroneous.
```

matriv commented on server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:
```diff
@@ -79,14 +80,12 @@ public ViewInfo create(RelationName ident, ClusterState state) {
+            analyzeError = true;
         }
-        return new ViewInfo(ident, view.stmt(), columns, view.owner());
+        String viewDefinition = analyzeError ? String.format(Locale.ENGLISH, "/* Corrupted view, needs fix */ %s", view.stmt()) : view.stmt();
```
How about adding a line break to make the comment even more visible.
```suggestion
        String viewDefinition = analyzeError ? String.format(Locale.ENGLISH, "/* Corrupted view, needs fix */\n%s", view.stmt()) : view.stmt();
```

matriv reviewed (commented):
Thank you! Left a couple of comments, also: could you please check if there is any possibility to have a unit test instead, by using the `ViewInfoFactory` (like for example in `DocIndexMetadataTest`)?

BaurzhanSakhariev commented:
> if there is any possibility to have a unit test instead

I added a unit test and removed Itest. I wanted to ensure that information schema is accessible but probably unit test is enough since we know that for that specific issue throwing an exception was the root cause.

matriv commented on server/src/test/java/io/crate/metadata/view/ViewInfoFactoryTest.java:
```diff
@@ -0,0 +1,57 @@
+        ClusterState state  = ClusterState.builder(ClusterState.EMPTY_STATE)
+            .metadata(Metadata.builder().putCustom(ViewsMetadata.TYPE, viewsMetadata)).build();
+
+        // Selecting from view itself throws an error.
```
Please adjust: 
```suggestion
        // `definition` col includes a hint about an error in the view's query
```

matriv reviewed (approved):
Thx!
